dell/xps/13-9310/default.nix

```diff
@@ -0,0 +1,89 @@
+{ lib, pkgs, ... }:
+{
+  imports = [
+    ../../../common/cpu/intel
+    ../../../common/pc/laptop
+  ];
+
+  # TODO: upstream to NixOS/nixpkgs
+  nixpkgs.overlays = [(final: previous: {
+    qca6390-firmware = final.callPackage ./qca6390-firmware.nix {};
+  })];
+
+  hardware.firmware = lib.mkBefore [
+    # Necessary for audio.
+    # https://bbs.archlinux.org/viewtopic.php?pid=1933548#p1933548
+    pkgs.sof-firmware
+    # Firmware for the AX500 (wi-fi & bluetooth chip).
+    pkgs.qca6390-firmware
+  ];
+
+  # Confirmed necessary to get audio working as of 2020-11-13:
+  # https://bbs.archlinux.org/viewtopic.php?pid=1933643#p1933643
+  boot.extraModprobeConfig = ''
+    options snd-intel-dspcfg dsp_driver=1
+  '';
+
+  # Touchpad goes over i2c.
+  # Without this we get errors in dmesg on boot and hangs when shutting down.
+  boot.blacklistedKernelModules = [ "psmouse" ];
+
+  # TODO: Remove this once landed in kernel.
+  # Apply kernel patch for xps 9310 wifi bug.
+  # https://patchwork.kernel.org/project/linux-wireless/patch/1605121102-14352-1-git-send-email-kvalo@codeaurora.org/
+  boot.kernelPatches = [
```

Comment:
With all these kernel patches in place we should probably set a specific kernel as they likely are no longer apply to a later one.

Reply:
Thanks for the review!

Currently kvalo's patches are based on `5.10-rc4`, but I can't see an "official" `linuxPackages_5_10_rc4` or anything like this in nixpkgs. I'm that instead I'll point directly `boot.kernelPackages` directly to [kvalo's branch](https://git.kernel.org/pub/scm/linux/kernel/git/kvalo/ath.git/log/?h=ath11k-qca6390-bringup) until all of these patches land in mainline. This would mean the nix expr will clone the whole kernel and building it locally though, but it's probably worth the improved reliability. I'll push an update for this soon, but let me know if there's a nicer way of achieving this.

Reply:
You can probably set `boot.kernelPackages` to the output of `buildLinux` with the [version set back to rc4](https://github.com/NixOS/nixpkgs/blob/bb15fa742780543ef87df2e6ea7e1747eadfc82b/pkgs/os-specific/linux/kernel/linux-testing.nix).

That said, they applied to rc5. _knocks wood_

Reply:
> You can probably set boot.kernelPackages to the output of buildLinux with the version set back to rc4.

Good idea! Do you know of an easier way to point to this version of `linuxPackages` while *also* applying kvalo's patches that doesn't involve linking a unique patch for each commit like I'm currently doing? E.g. perhaps there is some nicer way of specifying a single patch that links to the diff between kvalo's branch and rc4? I figure I could manually create and upload this diff somewhere, but thought that surely `git.kernel.org` must already provide some link for this that I'm missing?

Otherwise, I might just point to kvalo's branch to make maintaining this a bit easier, as they seem to be pushing new commits quite regularly.

Reply:
AFAIK, you're doing it the correct way. My understanding is kvalo's branch is tracking just ath-next, which may be out of sync with improvements to other parts of the kernel.